File: clojure/cars-assemble/src/cars_assemble.clj
(ns cars-assemble)

(defn success-rate [speed]
  (condp <= speed
    10 0.77
    9 0.8
    8 0.9
    4 1.0
    0 0.0
    ))

(defn production-rate-no-faulty [speed]
  (* 221 speed))

(defn production-rate
  "Returns the assembly line's production rate per hour,
   taking into account its success rate"
  [speed]
  (* (success-rate speed) (production-rate-no-faulty speed)))

(defn working-items
  "Calculates how many working cars are produced per minute"
  [speed]
  (int (quot (production-rate speed) 60)))

File: go/cars-assemble/cars_assemble.go
package cars

import "math"

// CalculateWorkingCarsPerHour calculates how many working cars are
// produced by the assembly line every hour.
func CalculateWorkingCarsPerHour(productionRate int, successRate float64) float64 {
	return float64(productionRate) * (successRate / 100)
}

// CalculateWorkingCarsPerMinute calculates how many working cars are
// produced by the assembly line every minute.
func CalculateWorkingCarsPerMinute(productionRate int, successRate float64) int {
	return int(CalculateWorkingCarsPerHour(productionRate, successRate) / 60)
}

// CalculateCost works out the cost of producing the given number of cars.
func CalculateCost(carsCount int) uint {
	ten_cars_cost := 95_000
	single_car_cost := 10_000

	ten_car_groups := int(math.Floor(float64(carsCount) / 10.0))
	remaining_single_cars := carsCount - (10 * ten_car_groups)

	return uint(ten_car_groups*ten_cars_cost + remaining_single_cars*(single_car_cost))
}
